### PR TITLE
Change internal representation of bucket key of time_series agg

### PR DESCRIPTION
**Pull request by martijnvg**
Currently, the key is a map, which can make reducing large response more memory intense then it should be also. Also data structures used during reduce are not back by bigarrays so not accounted for.

This commit changes how the key is represented internally. By using BytesRef instead of Map. This doesn't commit doesn't change how the key is represented in the response. It also changes the reduce method to make use of the bucket keys are now bytes refs.

Relates to #74660

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java [hunk 1]
```diff
@@ -186,38 +194,46 @@ protected void doWriteTo(StreamOutput out) throws IOException {
...
+        // TODO: optimize single result case either by having a if check here and return aggregations.get(0) or
+        // by overwriting the mustReduceOnSingleInternalAgg() method
+
+        final BigArrays bigArrays = reduceContext.bigArrays();
+        final int initialCapacity = aggregations.stream()
+            .map(value -> (InternalTimeSeries) value)
+            .mapToInt(value -> value.getBuckets().size())
+            .max()
+            .getAsInt();
+        try (LongObjectPagedHashMap<List<InternalBucket>> tsKeyToBuckets = new LongObjectPagedHashMap<>(initialCapacity, bigArrays)) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java:206]: Do we send these back in sorted order? Should we? That way we don't need to do this collection thing - we can just walk the results in parallel and marge.

File: server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesAggregator.java [hunk 2]
```diff
@@ -51,14 +50,14 @@ public TimeSeriesAggregator(
...
-            BytesRef spareKey = new BytesRef();
             BytesKeyedBucketOrds.BucketOrdsEnum ordsEnum = bucketOrds.ordsEnum(owningBucketOrds[ordIdx]);
             List<InternalTimeSeries.InternalBucket> buckets = new ArrayList<>();
             while (ordsEnum.next()) {
                 long docCount = bucketDocCount(ordsEnum.ord());
+                BytesRef spareKey = new BytesRef();
                 ordsEnum.readValue(spareKey);
                 InternalTimeSeries.InternalBucket bucket = new InternalTimeSeries.InternalBucket(
-                    TimeSeriesIdFieldMapper.decodeTsid(spareKey),
+                    spareKey,
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesAggregator.java:60]: I don't think it's correct to call it a "spare" any more if you are sending it back- it's just the `key`.

File: server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java:206]: Yes, I think the result is sorted. We collect the tsids in order in `BytesKeyedBucketOrds` and then we emit the tsids from `BytesKeyedBucketOrds` in the same way it was inserted.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java:206]: I think we probably should have a stronger guarantee - that the results are sorted by the bytes representation of the tsid. I think they already *are*. But we should, like, force it. Or at least `assert` it.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/timeseries/InternalTimeSeries.java]: I've added asserts for this: 0d20460c6a4d7b3bab9d288add6a80fff2343d73

File: modules/aggregations/src/main/java/org/elasticsearch/aggregations/bucket/timeseries/InternalTimeSeries.java [hunk 3]
```diff
@@ -187,38 +193,62 @@ protected void doWriteTo(StreamOutput out) throws IOException {
...
-        Map<Map<String, Object>, List<InternalBucket>> bucketsList = null;
+        // TODO: optimize single result case either by having a if check here and return aggregations.get(0) or
+        // by overwriting the mustReduceOnSingleInternalAgg() method
+        final int initialCapacity = aggregations.stream()
+            .map(value -> (InternalTimeSeries) value)
+            .mapToInt(value -> value.getBuckets().size())
+            .max()
+            .getAsInt();
+
+        final List<IteratorAndCurrent<InternalBucket>> iterators = new ArrayList<>(aggregations.size());
```
Comment (nik9000) [inline modules/aggregations/src/main/java/org/elasticsearch/aggregations/bucket/timeseries/InternalTimeSeries.java:204]: If this were a Lucene `PriorityQueue` I think you'd get for free finding the earliest tsid. As a bonus it'd have slightly better complexity as the number of shards goes up.

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by martijnvg** [comment]
Restarting CI builds, because of network issue:

```
Could not determine the dependencies of task ':x-pack:plugin:ml:explodedBundlePlugin'. |  
-- | --
  | [8.5.1] > Could not resolve all files for configuration ':x-pack:plugin:ml:nativeBundle'. |  
  | [8.5.1]    > Could not download ml-cpp-8.5.1-SNAPSHOT-deps.zip (org.elasticsearch.ml:ml-cpp:8.5.1-SNAPSHOT) |  
  | [8.5.1]       > Could not get resource 'https://artifacts-snapshot.elastic.co/ml-cpp/8.5.1-SNAPSHOT/downloads/ml-cpp/ml-cpp-8.5.1-SNAPSHOT-deps.zip'. |  
  | [8.5.1]          > Premature end of Content-Length delimited message body (expected: 340,418,397; received: 1,343,488)

```

**Review by nik9000 — APPROVED** [review]

**Pull request merged**